### ISSUE TITLE
atom-text-editor should size itself according to contents

### ISSUE THREAD

**Issue by jgebhardt** [atom/atom#12441]
### Description

It would be great if the `atom-text-editor` element sized itself according to its contents. We rely on this behavior in Nuclide to provide syntax highlighting inside [type-hints](https://nuclide.io/docs/languages/flow/#type-hinting), which uses an atom-text-editor inside a marker overlay. This used to work fine in Atom ≤1.8.x, although it relied on the `min-width: 1em` trick documented in the init script below. In 1.9.x, `.lines` elements inside the atom-text-editor shadow DOM don't have an explicit pixel width set anymore, which causes this behavior to break.

Ideally there would be an attribute that would make the text editor reliably auto-size itself according to its contents.

### Steps to Reproduce

Use the following init script (`init.js` in ~/.atom):
```es6
'use babel';

import {TextBuffer} from 'atom';

function setupTextEditor(props) {
  const textEditorParams = {
    buffer: props.textBuffer,
    lineNumberGutterVisible: false,
  };
  const textEditor = atom.workspace.buildTextEditor(textEditorParams);
  if (props.grammar != null) {
    textEditor.setGrammar(props.grammar);
  }
  return textEditor;
}

async function setup() {
  const buffer = new TextBuffer('const f = {a: 123};'); // dummy JS code
  const [jsGrammar] = atom.grammars.getGrammars().filter(g => g.name === 'JavaScript'); // hack to get JS grammar
  const textEditor = await atom.workspace.open();
  const range = textEditor.setText('test');
  const marker = textEditor.markBufferRange(range, {invalidate: 'never'});
  const datatipElement = document.createElement('div');
  const floatingEditor = setupTextEditor({
    grammar: jsGrammar,
    textBuffer: buffer,
  }); // floatingEditor is expected to size itself according to ints contents.
  datatipElement.style.border = '1px solid red';
  datatipElement.style.minWidth = '1em'; // hack to auto-size in 1.8.x and prior. Auto-sizing does not work in 1.9.x
  datatipElement.appendChild(floatingEditor.getElement());
  textEditor.decorateMarker(
    marker,
    {
      type: 'overlay',
      position: 'tail',
      item: datatipElement,
    },
  );
  return datatipElement;
}
atom.packages.onDidActivateInitialPackages(setup); // wait for grammar package to activate
```

**Expected behavior:**
1.8.x: The text-editor element (highlighted with red border) expands to fit its contents:
<img width="327" alt="screenshot 2016-08-16 18 28 15" src="https://cloud.githubusercontent.com/assets/632968/17721528/ad35a52e-63df-11e6-8ae3-03f6c8d9ee4c.png">

**Actual behavior:**
1.9.x: text-editor element does not set its width; it is `0` by default (1em in the screenshot due to applying the `min-width: 1em` trick):
<img width="455" alt="screenshot 2016-08-16 18 28 34" src="https://cloud.githubusercontent.com/assets/632968/17721526/a9aaff3a-63df-11e6-8c86-c8fabfdc8dfe.png">

### Versions
Broken in ≥ 1.9.x.

**Comment by nathansobo** [comment]
This regression was introduced in https://github.com/atom/atom/pull/11866, which itself was needed to fix https://github.com/atom/find-and-replace/issues/732, a layout regression introduced by an Electron upgrade. I *think* the original issue was caused by a change in flexbox behavior in Chrome.

The big question in my mind is whether we could fix #732 in a way that preserves the original auto-width behavior of the editor, or whether we should plow forward with explicit properties to control `autoWidth` and `autoHeight`.

We currently attempt to automatically determine whether to auto-height the editor based on DOM properties, but it's actually pretty janky and I don't think we have the kind of low-level access to the layout logic to make this decision perfectly. So part of me wishes that auto-height and auto-width were *both* only enabled via explicit properties. However, if we were to do that we would probably break current usages of the editor that rely on the automatic enabling of auto-height.

So, if we make the auto-width property explicit, there will be some inconsistency with auto-height being implicit. Perhaps that's okay. We could honor any explicit `autoHeight` property and otherwise continue to auto-detect to keep things working.

@as-cii what do you think? Also cc @maxbrunsfeld.

**Comment by jgebhardt** [comment]
W00t! Thanks for the quick fix @nathansobo!
